receiver/awscontainerinsightreceiver/internal/stores/podstore.go

```diff
@@ -0,0 +1,606 @@
+		prefFullPodName:  prefFullPodName,
+	}
+
+	// Try to detect kubelet permission issue here
+	if _, err := podStore.kubeletClient.ListPods(); err != nil {
+		return nil, fmt.Errorf("cannot get pod from kubelet, err: %v", err)
+	}
+
+	return podStore, nil
+}
+
+func (p *PodStore) getPrevMeasurement(metricType, metricKey string) (interface{}, bool) {
+	prevMeasurement, ok := p.prevMeasurements[metricType]
+	if !ok {
+		return nil, false
+	}
+
+	content, ok := prevMeasurement.Get(metricKey)
+
+	if !ok {
+		return nil, false
+	}
+
+	return content, true
+}
+
+func (p *PodStore) setPrevMeasurement(metricType, metricKey string, content interface{}) {
+	prevMeasurement, ok := p.prevMeasurements[metricType]
+	if !ok {
+		prevMeasurement = newMapWithExpiry(measurementsExpiry)
+		p.prevMeasurements[metricType] = prevMeasurement
+	}
+	prevMeasurement.Set(metricKey, content)
+}
+
+func (p *PodStore) RefreshTick() {
+	now := time.Now()
+	if now.Sub(p.lastRefreshed) >= refreshInterval {
+		p.refresh(now)
+		// call cleanup every refresh cycle
+		p.cleanup(now)
+		p.lastRefreshed = now
+	}
+}
+
+func (p *PodStore) Decorate(metric CIMetric, kubernetesBlob map[string]interface{}) bool {
+	if metric.GetTag(ci.MetricType) == ci.TypeNode {
+		p.decorateNode(metric)
+	} else if metric.GetTag(ci.K8sPodNameKey) != "" {
+		podKey := createPodKeyFromMetric(metric)
+		if podKey == "" {
+			p.logger.Error("podKey is unavailable when decorating pod")
+			return false
+		}
+
+		entry := p.getCachedEntry(podKey)
+		if entry == nil {
+			p.logger.Debug(fmt.Sprintf("no pod is found for %s, refresh the cache now...", podKey))
+			p.refresh(time.Now())
+			entry = p.getCachedEntry(podKey)
+		}
+
+		// If pod is still not found, insert a placeholder to avoid too many refresh
+		if entry == nil {
+			log.Printf("W! no pod is found after reading through kubelet, add a placeholder for %s", podKey)
+			p.setCachedEntry(podKey, &cachedEntry{creation: time.Now()})
+			return false
+		}
+
+		// If the entry is not a placeholder, decorate the pod
+		if entry.pod.Name != "" {
+			p.decorateCPU(metric, &entry.pod)
+			p.decorateMem(metric, &entry.pod)
+			p.addStatus(metric, &entry.pod)
+			addContainerCount(metric, &entry.pod)
+			addContainerID(&entry.pod, metric, kubernetesBlob, p.logger)
+			p.addPodOwnersAndPodName(metric, &entry.pod, kubernetesBlob)
+			addLabels(&entry.pod, kubernetesBlob)
+		} else {
+			p.logger.Warn(fmt.Sprintf("no pod information is found in podstore for pod %s", podKey))
+			return false
+		}
+	}
+	return true
+}
+
+func (p *PodStore) getCachedEntry(podKey string) *cachedEntry {
+	p.Lock()
+	defer p.Unlock()
+	if content, ok := p.cache.Get(podKey); ok {
+		return content.(*cachedEntry)
+	}
+	return nil
+}
+
+func (p *PodStore) setCachedEntry(podKey string, entry *cachedEntry) {
+	p.Lock()
+	defer p.Unlock()
+	p.cache.Set(podKey, entry)
+}
+
+func (p *PodStore) setNodeStats(stats nodeStats) {
+	p.Lock()
+	defer p.Unlock()
+	p.nodeInfo.nodeStats = stats
```

Comment:
Reaching directly into `nodeInfo` like this doesn't feel right.  There should probably be accessors/mutators on it for the values in `nodeStats` that lock appropriately.